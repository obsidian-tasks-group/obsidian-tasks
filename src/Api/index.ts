import type { App } from 'obsidian';
import { createTaskLineModal } from './createTaskLineModal';
import type { TasksApiV1 } from './TasksApiV1';
import { defaultTaskModalFactory } from './createTaskLineModalHelper';

/**
 * Factory method for API v1
 *
 * @param app - The Obsidian App
 */
export const tasksApiV1 = (app: App): TasksApiV1 => {
    return {
        createTaskLineModal: (): Promise<string> => {
            return createTaskLineModal(app, defaultTaskModalFactory);
        },
    };
};
