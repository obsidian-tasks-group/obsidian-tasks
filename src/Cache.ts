import { MetadataCache, TAbstractFile, TFile, Vault } from 'obsidian';
import type { CachedMetadata, EventRef } from 'obsidian';
import type { HeadingCache, ListItemCache, SectionCache } from 'obsidian';
import { Mutex } from 'async-mutex';

import { Task } from './Task';
import type { Events } from './Events';

export enum State {
    Cold = 'Cold',
    Initializing = 'Initializing',
    Warm = 'Warm',
}

export class Cache {
    private readonly metadataCache: MetadataCache;
    private readonly metadataCacheEventReferences: EventRef[];
    private readonly vault: Vault;
    private readonly vaultEventReferences: EventRef[];
    private readonly events: Events;
    private readonly eventsEventReferences: EventRef[];

    private readonly tasksMutex: Mutex;
    private state: State;
    private tasks: Task[];

    /**
     * We cannot know if this class will be instantiated because obsidian started
     * or because the plugin was activated later. This means we have to load the
     * whole vault once after the first metadata cache resolve to ensure that we
     * load the entire vault in case obsidian is starting up. In the case of
     * obsidian starting, the task cache's initial load would end up with 0 tasks,
     * as the metadata cache would still be empty.
     */
    private loadedAfterFirstResolve: boolean;

    constructor({
        metadataCache,
        vault,
        events,
    }: {
        metadataCache: MetadataCache;
        vault: Vault;
        events: Events;
    }) {
        this.metadataCache = metadataCache;
        this.metadataCacheEventReferences = [];
        this.vault = vault;
        this.vaultEventReferences = [];
        this.events = events;
        this.eventsEventReferences = [];

        this.tasksMutex = new Mutex();
        this.state = State.Cold;
        this.tasks = [];

        this.loadedAfterFirstResolve = false;

        this.subscribeToCache();
        this.subscribeToVault();
        this.subscribeToEvents();

        this.loadVault();
    }

    public unload(): void {
        for (const eventReference of this.metadataCacheEventReferences) {
            this.metadataCache.offref(eventReference);
        }

        for (const eventReference of this.vaultEventReferences) {
            this.vault.offref(eventReference);
        }

        for (const eventReference of this.eventsEventReferences) {
            this.events.off(eventReference);
        }
    }

    public getTasks(): Task[] {
        return this.tasks;
    }

    public getState(): State {
        return this.state;
    }

    private notifySubscribers(): void {
        this.events.triggerCacheUpdate({
            tasks: this.tasks,
            state: this.state,
        });
    }

    private subscribeToCache(): void {
        const resolvedEventeReference = this.metadataCache.on(
            'resolved',
            async () => {
                // Resolved fires on every change.
                // We only want to initialize if we haven't already.
                if (!this.loadedAfterFirstResolve) {
                    this.loadedAfterFirstResolve = true;
                    this.loadVault();
                }
            },
        );
        this.metadataCacheEventReferences.push(resolvedEventeReference);

        // Does not fire when starting up obsidian and only works for changes.
        const changedEventReference = this.metadataCache.on(
            'changed',
            (file: TFile) => {
                this.tasksMutex.runExclusive(() => {
                    this.indexFile(file);
                });
            },
        );
        this.metadataCacheEventReferences.push(changedEventReference);
    }

    private subscribeToVault(): void {
        const createdEventReference = this.vault.on(
            'create',
            (file: TAbstractFile) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.indexFile(file);
                });
            },
        );
        this.vaultEventReferences.push(createdEventReference);

        const deletedEventReference = this.vault.on(
            'delete',
            (file: TAbstractFile) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.tasks = this.tasks.filter((task: Task) => {
                        return task.path !== file.path;
                    });

                    this.notifySubscribers();
                });
            },
        );
        this.vaultEventReferences.push(deletedEventReference);

        const renamedEventReference = this.vault.on(
            'rename',
            (file: TAbstractFile, oldPath: string) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.tasks = this.tasks.map((task: Task): Task => {
                        if (task.path === oldPath) {
                            return new Task({ ...task, path: file.path });
                        } else {
                            return task;
                        }
                    });

                    this.notifySubscribers();
                });
            },
        );
        this.vaultEventReferences.push(renamedEventReference);
    }

    private subscribeToEvents(): void {
        const requestReference = this.events.onRequestCacheUpdate((handler) => {
            handler({ tasks: this.tasks, state: this.state });
        });
        this.eventsEventReferences.push(requestReference);
    }

    private loadVault(): Promise<void> {
        return this.tasksMutex.runExclusive(async () => {
            this.state = State.Initializing;
            await Promise.all(
                this.vault.getMarkdownFiles().map((file: TFile) => {
                    return this.indexFile(file);
                }),
            );
            this.state = State.Warm;
            // Notify that the cache is now warm:
            this.notifySubscribers();
        });
    }

    private async indexFile(file: TFile): Promise<void> {
        const fileCache = this.metadataCache.getFileCache(file);
        if (fileCache === null || fileCache === undefined) {
            return;
        }

        const oldTasks = this.tasks.filter((task: Task) => {
            return task.path === file.path;
        });

        const listItems = fileCache.listItems;
        // When there is no list items cache, there are no tasks.
        // Still continue to notify watchers of removal.

        let newTasks: Task[] = [];
        if (listItems !== undefined) {
            // Only read the file and process for tasks if there are list items.
            const fileContent = await this.vault.cachedRead(file);
            newTasks = Cache.getTasksFromFileContent(
                fileContent,
                listItems,
                fileCache,
                file,
            );
        }

        // If there are no changes in any of the tasks, there's
        // nothing to do, so just return.
        if (Task.tasksListsIdentical(oldTasks, newTasks)) {
            // This code kept for now, to allow for debugging during development.
            // It is too verbose to release to users.
            // if (this.getState() == State.Warm) {
            //     console.debug(`Tasks unchanged in ${file.path}`);
            // }
            return;
        }

        if (this.getState() == State.Warm) {
            console.debug(
                `At least one task, its line number or its heading has changed in ${file.path}: triggering a refresh of all active Tasks blocks in Live Preview and Reading mode views.`,
            );
        }

        // Remove all tasks from this file from the cache before
        // adding the ones that are currently in the file.
        this.tasks = this.tasks.filter((task: Task) => {
            return task.path !== file.path;
        });

        this.tasks.push(...newTasks);

        // All updated, inform our subscribers.
        this.notifySubscribers();
    }

    private static getTasksFromFileContent(
        fileContent: string,
        listItems: ListItemCache[],
        fileCache: CachedMetadata,
        file: TFile,
    ): Task[] {
        const tasks: Task[] = [];
        const fileLines = fileContent.split('\n');

        // We want to store section information with every task so
        // that we can use that when we post process the markdown
        // rendered lists.
        let currentSection: SectionCache | null = null;
        let sectionIndex = 0;
        for (const listItem of listItems) {
            if (listItem.task !== undefined) {
                if (
                    currentSection === null ||
                    currentSection.position.end.line <
                        listItem.position.start.line
                ) {
                    // We went past the current section (or this is the first task).
                    // Find the section that is relevant for this task and the following of the same section.
                    currentSection = Cache.getSection(
                        listItem.position.start.line,
                        fileCache.sections,
                    );
                    sectionIndex = 0;
                }

                if (currentSection === null) {
                    // Cannot process a task without a section.
                    continue;
                }

                const line = fileLines[listItem.position.start.line];
                const task = Task.fromLine({
                    line,
                    path: file.path,
                    sectionStart: currentSection.position.start.line,
                    sectionIndex,
                    precedingHeader: Cache.getPrecedingHeader(
                        listItem.position.start.line,
                        fileCache.headings,
                    ),
                });

                if (task !== null) {
                    sectionIndex++;
                    tasks.push(task);
                }
            }
        }

        return tasks;
    }

    private static getSection(
        lineNumberTask: number,
        sections: SectionCache[] | undefined,
    ): SectionCache | null {
        if (sections === undefined) {
            return null;
        }

        for (const section of sections) {
            if (
                section.position.start.line <= lineNumberTask &&
                section.position.end.line >= lineNumberTask
            ) {
                return section;
            }
        }

        return null;
    }

    private static getPrecedingHeader(
        lineNumberTask: number,
        headings: HeadingCache[] | undefined,
    ): string | null {
        if (headings === undefined) {
            return null;
        }

        let precedingHeader: string | null = null;

        for (const heading of headings) {
            if (heading.position.start.line > lineNumberTask) {
                return precedingHeader;
            }
            precedingHeader = heading.heading;
        }
        return precedingHeader;
    }
}
