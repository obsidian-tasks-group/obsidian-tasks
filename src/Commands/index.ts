import type { App, Editor, MarkdownFileInfo, MarkdownView, View } from 'obsidian';
import type TasksPlugin from '../main';
import { createOrEdit } from './CreateOrEdit';

import { toggleDone } from './ToggleDone';

export class Commands {
    private readonly plugin: TasksPlugin;

    private get app(): App {
        return this.plugin.app;
    }

    constructor({ plugin }: { plugin: TasksPlugin }) {
        this.plugin = plugin;

        plugin.addCommand({
            id: 'edit-task',
            name: 'Create or edit task',
            icon: 'pencil',
            editorCheckCallback: (checking: boolean, editor: Editor, view: MarkdownView | MarkdownFileInfo) => {
                // TODO Need to explore what happens if a tasks code block is rendered before the Cache has been created.
                return createOrEdit(checking, editor, view as View, this.app, this.plugin.getTasks());
            },
        });

        plugin.addCommand({
            id: 'toggle-done',
            name: 'Toggle task done',
            icon: 'check-in-circle',
            editorCheckCallback: toggleDone,
        });
    }
}
