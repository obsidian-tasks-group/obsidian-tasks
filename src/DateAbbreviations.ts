// Abbreviations for entering dates with chrono
// MAINTENANCE NOTE:
//      If adding more abbreviations, please review datePlaceholder in src/ui/EditTask.svelte
const abbreviations = {
    td: 'today',
    tm: 'tomorrow',
    yd: 'yesterday',
    tw: 'this week',
    nw: 'next week',
    weekend: 'sat',
    we: 'sat',
};

/**
 * Expand any recognised abbreviations for dates.
 *
 * Important: the abbreviation is only expanded if it is foolowed by a space.
 *
 * For example, 'td ' is expanded to 'today'
 * @param date
 */
export function doAutocomplete(date: string): string {
    for (const [key, val] of Object.entries(abbreviations)) {
        date = date.replace(RegExp(`\\b${key}\\s`, 'i'), val);
    }
    return date;
}
