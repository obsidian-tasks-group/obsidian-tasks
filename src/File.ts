import { MetadataCache, TFile, Vault } from 'obsidian';
import type { ListItemCache } from 'obsidian';

import { getSettings } from './Config/Settings';
import type { Task } from './Task';

let metadataCache: MetadataCache | undefined;
let vault: Vault | undefined;

export const initializeFile = ({
    metadataCache: newMetadataCache,
    vault: newVault,
}: {
    metadataCache: MetadataCache;
    vault: Vault;
}) => {
    metadataCache = newMetadataCache;
    vault = newVault;
};

/**
 * Replaces the original task with one or more new tasks.
 *
 * If you pass more than one replacement task, all subsequent tasks in the same
 * section must be re-rendered, as their section indexes change. Assuming that
 * this is done faster than user interaction in practice.
 */
export const replaceTaskWithTasks = async ({
    originalTask,
    newTasks,
}: {
    originalTask: Task;
    newTasks: Task | Task[];
}): Promise<void> => {
    if (vault === undefined || metadataCache === undefined) {
        console.error('Tasks: cannot use File before initializing it.');
        return;
    }

    if (!Array.isArray(newTasks)) {
        newTasks = [newTasks];
    }

    tryRepetitive({
        originalTask,
        newTasks,
        vault,
        metadataCache,
        previousTries: 0,
    });
};

/**
 * This is a workaround to re-try when the returned file cache is `undefined`.
 * Retrying after a while may return a valid file cache.
 * Reported in https://github.com/obsidian-tasks-group/obsidian-tasks/issues/87
 */
const tryRepetitive = async ({
    originalTask,
    newTasks,
    vault,
    metadataCache,
    previousTries,
}: {
    originalTask: Task;
    newTasks: Task[];
    vault: Vault;
    metadataCache: MetadataCache;
    previousTries: number;
}): Promise<void> => {
    const retry = () => {
        if (previousTries > 10) {
            console.error('Tasks: Too many retries. File update not possible ...');
            return;
        }

        const timeout = Math.min(Math.pow(10, previousTries), 100); // 1, 10, 100, 100, 100, ...
        setTimeout(() => {
            tryRepetitive({
                originalTask,
                newTasks,
                vault,
                metadataCache,
                previousTries: previousTries + 1,
            });
        }, timeout);
    };

    const file = vault.getAbstractFileByPath(originalTask.path);
    if (!(file instanceof TFile)) {
        console.warn(`Tasks: No file found for task ${originalTask.description}. Retrying ...`);
        return retry();
    }

    if (file.extension !== 'md') {
        console.error('Tasks: Only supporting files with the .md file extension.');
        return;
    }

    const fileCache = metadataCache.getFileCache(file);
    if (fileCache == undefined || fileCache === null) {
        console.warn(`Tasks: No file cache found for file ${file.path}. Retrying ...`);
        return retry();
    }

    const listItemsCache = fileCache.listItems;
    if (listItemsCache === undefined || listItemsCache.length === 0) {
        console.warn(`Tasks: No list items found in file cache of ${file.path}. Retrying ...`);
        return retry();
    }

    const fileContent = await vault.read(file);
    const fileLines = fileContent.split('\n');

    const { globalFilter } = getSettings();
    let listItem: ListItemCache | undefined;
    let sectionIndex = 0;
    for (const listItemCache of listItemsCache) {
        if (listItemCache.position.start.line < originalTask.sectionStart) {
            continue;
        }

        if (listItemCache.task === undefined) {
            continue;
        }

        const line = fileLines[listItemCache.position.start.line];

        if (line.includes(globalFilter)) {
            if (sectionIndex === originalTask.sectionIndex) {
                listItem = listItemCache;
                break;
            }

            sectionIndex++;
        }
    }
    if (listItem === undefined) {
        console.error('Tasks: could not find task to toggle in the file.');
        return;
    }

    const updatedFileLines = [
        ...fileLines.slice(0, listItem.position.start.line),
        ...newTasks.map((task: Task) => task.toFileLineString()),
        ...fileLines.slice(listItem.position.start.line + 1), // Only supports single-line tasks.
    ];

    await vault.modify(file, updatedFileLines.join('\n'));
};
