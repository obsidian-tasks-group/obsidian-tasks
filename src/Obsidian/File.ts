import { type ListItemCache, MetadataCache, Notice, TFile, Vault, Workspace } from 'obsidian';
import { GlobalFilter } from '../Config/GlobalFilter';
import { type MockListItemCache, type MockTask, saveMockDataForTesting } from '../lib/MockDataCreator';
import type { Task } from '../Task/Task';
import { logging } from '../lib/logging';
import { logEndOfTaskEdit, logStartOfTaskEdit } from '../lib/LogTasksHelper';

let metadataCache: MetadataCache | undefined;
let vault: Vault | undefined;
let workspace: Workspace | undefined;

const supportedFileExtensions = ['md'];

function getFileLogger() {
    // For logging to actually produce debug output when enabled in settings,
    // it appears that the logger cannot be created until execution time.
    return logging.getLogger('tasks.File');
}

export type ErrorLoggingFunction = (message: string) => void;

export const initializeFile = ({
    metadataCache: newMetadataCache,
    vault: newVault,
    workspace: newWorkspace,
}: {
    metadataCache: MetadataCache;
    vault: Vault;
    workspace: Workspace;
}) => {
    metadataCache = newMetadataCache;
    vault = newVault;
    workspace = newWorkspace;
};

/**
 * Replaces the original task with one or more new tasks.
 *
 * If you pass more than one replacement task, all subsequent tasks in the same
 * section must be re-rendered, as their section indexes change. Assuming that
 * this is done faster than user interaction in practice.
 *
 * In addition, this function is meant to be called with reasonable confidence
 * that the {@code originalTask} is unmodified and at the exact same section and
 * sectionIdx in the source file it was originally found in. It will fail otherwise.
 */
export const replaceTaskWithTasks = async ({
    originalTask,
    newTasks,
}: {
    originalTask: Task;
    newTasks: Task | Task[];
}): Promise<void> => {
    if (vault === undefined || metadataCache === undefined || workspace === undefined) {
        errorAndNotice('Tasks: cannot use File before initializing it.');
        return;
    }

    if (!Array.isArray(newTasks)) {
        newTasks = [newTasks];
    }

    const logger = getFileLogger();
    const codeLocation = 'replaceTaskWithTasks()';
    logStartOfTaskEdit(logger, codeLocation, originalTask);
    logEndOfTaskEdit(logger, codeLocation, newTasks);

    await tryRepetitive({
        originalTask,
        newTasks,
        vault,
        metadataCache,
        workspace,
        previousTries: 0,
    });
};

/**
 * @todo Unify this with {@link showError} in EditorSuggestorPopup.ts
 * @param message
 */
function errorAndNotice(message: string) {
    console.error(message);
    new Notice(message, 15000);
}

function warnAndNotice(message: string) {
    console.warn(message);
    new Notice(message, 10000);
}

function debugLog(message: string) {
    const logger = getFileLogger();
    logger.debug(message);
}

// When this exception is thrown, it is meant to indicate that the caller should consider to try the operation
// again soon
class WarningWorthRetrying extends Error {}
// Same as above, but be silent about it
class RetryWithoutWarning extends Error {}

/**
 * This is a workaround to re-try when the returned file cache is `undefined`.
 * Retrying after a while may return a valid file cache.
 * Reported in https://github.com/obsidian-tasks-group/obsidian-tasks/issues/87
 */
const tryRepetitive = async ({
    originalTask,
    newTasks,
    vault,
    metadataCache,
    workspace,
    previousTries,
}: {
    originalTask: Task;
    newTasks: Task[];
    vault: Vault;
    metadataCache: MetadataCache;
    workspace: Workspace;
    previousTries: number;
}): Promise<void> => {
    const logger = getFileLogger();
    logger.debug(`tryRepetitive after ${previousTries} previous tries`);
    const retry = async () => {
        if (previousTries > 10) {
            const message = `Tasks: Could not find the correct task line to update.

The task line not updated is:
${originalTask.originalMarkdown}

In this markdown file:
"${originalTask.taskLocation.path}"

Note: further clicks on this checkbox will usually now be ignored until the file is opened (or certain, specific edits are made - it's complicated).

Recommendations:

1. Close all panes that have the above file open, and then re-open the file.

2. Check for exactly identical copies of the task line, in this file, and see if you can make them different.
`;
            errorAndNotice(message);
            return;
        }

        const timeout = Math.min(Math.pow(10, previousTries), 100); // 1, 10, 100, 100, 100, ...
        logger.debug(`timeout = ${timeout}`);
        setTimeout(async () => {
            await tryRepetitive({
                originalTask,
                newTasks,
                vault,
                metadataCache,
                workspace,
                previousTries: previousTries + 1,
            });
        }, timeout);
    };

    try {
        const [taskLineNumber, file, fileLines] = await getTaskAndFileLines(originalTask, vault);
        // Finally, we can insert 1 or more lines over the original task line:
        const updatedFileLines = [
            ...fileLines.slice(0, taskLineNumber),
            ...newTasks.map((task: Task) => task.toFileLineString()),
            ...fileLines.slice(taskLineNumber + 1), // Only supports single-line tasks.
        ];

        await vault.modify(file, updatedFileLines.join('\n'));
    } catch (e) {
        if (e instanceof WarningWorthRetrying) {
            if (e.message) warnAndNotice(e.message);
            await retry();
            return;
        } else if (e instanceof RetryWithoutWarning) {
            await retry();
            return;
        } else if (e instanceof Error) {
            errorAndNotice(e.message);
        }
    }
};

/*
 * This method returns the line on which `task` is defined, together with the file it is defined in, and the
 * lines of that file, possibly for the purpose of updating the task in the file or jumping to it.
 * It may throw a WarningWorthRetrying exception in several cases that justify a retry (should be handled by the caller)
 * or an Error exception in case of an unrecoverable error.
 */
async function getTaskAndFileLines(task: Task, vault: Vault): Promise<[number, TFile, string[]]> {
    if (metadataCache === undefined) throw new WarningWorthRetrying();
    // Validate our inputs.
    // For permanent failures, return nothing.
    // For failures that might be fixed if we wait for a little while, return retry().
    const file = vault.getAbstractFileByPath(task.path);
    if (!(file instanceof TFile)) {
        throw new WarningWorthRetrying(`Tasks: No file found for task ${task.description}. Retrying ...`);
    }

    if (!supportedFileExtensions.includes(file.extension)) {
        throw new Error(`Tasks: Does not support files with the ${file.extension} file extension.`);
    }

    const fileCache = metadataCache.getFileCache(file);
    if (fileCache == undefined || fileCache === null) {
        throw new WarningWorthRetrying(`Tasks: No file cache found for file ${file.path}. Retrying ...`);
    }

    const listItemsCache = fileCache.listItems;
    if (listItemsCache === undefined || listItemsCache.length === 0) {
        throw new WarningWorthRetrying(`Tasks: No list items found in file cache of ${file.path}. Retrying ...`);
    }

    // We can now try and find which line in the file currently contains originalTask,
    // so that we know which line to update.
    const fileContent = await vault.read(file); // TODO: replace with vault.process.
    const fileLines = fileContent.split('\n');
    const taskLineNumber = findLineNumberOfTaskToToggle(task, fileLines, listItemsCache, debugLog);

    if (taskLineNumber === undefined) {
        const logDataForMocking = false;
        if (logDataForMocking) {
            // There was an error finding the correct line to toggle,
            // so write out to the console a representation of the data needed to reconstruct the above
            // findLineNumberOfTaskToToggle() call, so that the content can be saved
            // to a JSON file and then re-used in a 'unit' test.
            saveMockDataForTesting(task, fileLines, listItemsCache);
        }
        throw new RetryWithoutWarning();
    }
    return [taskLineNumber, file, fileLines];
}

// A simpler version of the method above, which doesn't return the lines of the file, and handles exceptions
// internally rather than throw them to the outside
export async function getTaskLineAndFile(task: Task, vault: Vault): Promise<[number, TFile] | undefined> {
    try {
        const [taskLineNumber, file, _] = await getTaskAndFileLines(task, vault);
        return [taskLineNumber, file];
    } catch (e) {
        if (e instanceof WarningWorthRetrying) {
            if (e.message) warnAndNotice(e.message);
        } else if (e instanceof Error) {
            errorAndNotice(e.message);
        }
    }
    return undefined;
}

function isValidLineNumber(listItemLineNumber: number, fileLines: string[]) {
    return listItemLineNumber < fileLines.length;
}

/**
 * Try to find the line number of the originalTask
 * @param originalTask - the {@link Task} line that the user clicked on
 * @param fileLines - the lines read from the file.
 * @param listItemsCache
 * @param errorLoggingFunction - a function of type {@link ErrorLoggingFunction} - which will be called if the found
 *                               line differs from the original markdown in {@link originalTask}.
 *                               This parameter is provided to allow tests to be written for this code
 *                               that do not display a popup warning, but instead capture the error message.
 */
export function findLineNumberOfTaskToToggle(
    originalTask: Task | MockTask,
    fileLines: string[],
    listItemsCache: ListItemCache[] | MockListItemCache[],
    errorLoggingFunction: ErrorLoggingFunction,
): number | undefined {
    let result: number | undefined = tryFindingExactMatchAtOriginalLineNumber(originalTask, fileLines);
    if (result !== undefined) {
        return result;
    }

    result = tryFindingIdenticalUniqueMarkdownLineInFile(originalTask, fileLines);
    if (result !== undefined) {
        return result;
    }

    return tryFindingLineNumberFromTaskSectionInfo(originalTask, fileLines, listItemsCache, errorLoggingFunction);
}

/**
 *  If the line at line number in originalTask matches original markdown,
 *  treat that as the correct answer.
 *
 *  This could go wrong if:
 *     - Some lines have been added since originalTask was rendered in Reading view,
 *       and an identical task line was added, that happened by coincidence to be in the same
 *       line number as the original task.
 *
 * @param originalTask
 * @param fileLines
 */
function tryFindingExactMatchAtOriginalLineNumber(originalTask: Task | MockTask, fileLines: string[]) {
    const originalTaskLineNumber = originalTask.taskLocation.lineNumber;
    if (isValidLineNumber(originalTaskLineNumber, fileLines)) {
        if (fileLines[originalTaskLineNumber] === originalTask.originalMarkdown) {
            const logger = getFileLogger();
            logger.debug(`Found original markdown at original line number ${originalTaskLineNumber}`);
            return originalTaskLineNumber;
        }
    }
    return undefined;
}

/**
 * If the line only appears once in the file, use that line number.
 *
 * This could go wrong if:
 *    - the user had commented out the original task line, and the section had not yet been redrawn
 * @param originalTask
 * @param fileLines
 */
function tryFindingIdenticalUniqueMarkdownLineInFile(originalTask: Task | MockTask, fileLines: string[]) {
    const matchingLineNumbers = [];
    for (let i = 0; i < fileLines.length; i++) {
        if (fileLines[i] === originalTask.originalMarkdown) {
            matchingLineNumbers.push(i);
        }
    }
    if (matchingLineNumbers.length === 1) {
        // There is only one instance of the line in the file, so it must be the
        // line we are looking for.
        return matchingLineNumbers[0];
    }
    return undefined;
}

/**
 * Fall back on the original algorithm, which uses the section information inside the task's {@link TaskLocation}.
 *
 * @param originalTask
 * @param fileLines
 * @param listItemsCache
 * @param errorLoggingFunction
 */
function tryFindingLineNumberFromTaskSectionInfo(
    originalTask: Task | MockTask,
    fileLines: string[],
    listItemsCache: ListItemCache[] | MockListItemCache[],
    errorLoggingFunction: ErrorLoggingFunction,
) {
    let taskLineNumber: number | undefined;
    let sectionIndex = 0;
    for (const listItemCache of listItemsCache) {
        const listItemLineNumber = listItemCache.position.start.line;
        if (!isValidLineNumber(listItemLineNumber, fileLines)) {
            // One or more lines has been deleted since the cache was populated,
            // so there is at least one list item in the cache that is beyond
            // the end of the actual file on disk.
            return undefined;
        }

        if (listItemLineNumber < originalTask.taskLocation.sectionStart) {
            continue;
        }

        if (listItemCache.task === undefined) {
            continue;
        }

        const line = fileLines[listItemLineNumber];
        if (GlobalFilter.getInstance().includedIn(line)) {
            if (sectionIndex === originalTask.taskLocation.sectionIndex) {
                if (line === originalTask.originalMarkdown) {
                    taskLineNumber = listItemLineNumber;
                } else {
                    errorLoggingFunction(
                        `Tasks: Unable to find task in file ${originalTask.taskLocation.path}.
Expected task:
${originalTask.originalMarkdown}
Found task:
${line}`,
                    );
                    return;
                }
                break;
            }

            sectionIndex++;
        }
    }
    return taskLineNumber;
}
