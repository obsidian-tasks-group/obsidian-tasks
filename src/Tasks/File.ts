import { Obsidian } from 'Obsidian';
import { Task } from './Task';

export class File {
    private readonly obsidian: Obsidian;

    constructor({ obsidian }: { obsidian: Obsidian }) {
        this.obsidian = obsidian;
    }

    public async toggleDone({
        path,
        lineNumber,
    }: {
        path: string;
        lineNumber: number;
    }): Promise<void> {
        const lines = await this.obsidian.readLines({ path });
        if (lines === undefined) {
            return;
        }

        const line = lines[lineNumber];
        lines[lineNumber] = this.toggleLine({
            line,
            path,
            lineNumber,
        });

        await this.obsidian.writeLines({ path, lines });
    }

    private toggleLine({
        line,
        path,
        lineNumber,
    }: {
        line: string;
        path: string;
        lineNumber: number;
    }): string {
        const task = Task.fromLine({
            line,
            path,
            lineNumber,
            pageIndex: undefined,
            precedingHeader: undefined,
        });
        if (task === undefined) {
            return line;
        }

        const toggledTask = task.toggle();

        return toggledTask.toFileString();
    }
}
