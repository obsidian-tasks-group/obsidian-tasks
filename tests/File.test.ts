import { readFileSync } from 'fs';
import { verify } from 'approvals/lib/Providers/Jest/JestApprovals';
import { findLineNumberOfTaskToToggle } from '../src/File';
import type { MockTogglingDataForTesting } from '../src/lib/MockDataCreator';

/**
 * A function to help test File.findLineNumberOfTaskToToggle()
 *
 * It uses mock data to simulate user behaviour.
 *
 * This is a fragile method of testing, but right now is better than no automated
 * testing at all.
 *
 * Input files for {@link jsonFileName} are in `tests/__test_data__/MockDataForTogglingTasks`
 * They are created by:
 *   1. setting logDataForMocking to true in File.ts.
 *   2. Running a local build on the plugin
 *   3. Showing the console
 *   4. Setting up a small markdown file with the scenario to test
 *   5. Working the steps required to make the toggle operation fail
 *   6. Grabbing the console output
 *   7. Pasting it in to a new JSON file in `tests/__test_data__/MockDataForTogglingTasks`
 *   8. Auto-formatting the JSON so that it is readable.
 * @param jsonFileName
 * @param taskLineToToggle
 * @param expectedLineNumber
 * @param actualIncorrectLineFound - optional, only needed if the wrong line would be toggled.
 */
function testFindLineNumberOfTaskToToggle(
    jsonFileName: string,
    taskLineToToggle: string,
    expectedLineNumber: number | undefined,
    actualIncorrectLineFound?: string,
) {
    // Arrange
    const data = readFileSync('tests/__test_data__/MockDataForTogglingTasks/' + jsonFileName, 'utf-8');
    const everything: MockTogglingDataForTesting = JSON.parse(data);
    expect(everything.taskData.originalMarkdown).toEqual(taskLineToToggle);

    // Act
    const originalTask = everything.taskData;
    const fileLines = everything.fileData.fileLines;
    const listItemsCache = everything.cacheData.listItemsCache;

    let errorString: string | undefined;
    const captureError = (message: string) => {
        errorString = message;
    };

    const result = findLineNumberOfTaskToToggle(
        originalTask,
        fileLines,
        listItemsCache,
        captureError.bind(errorString),
    );

    // Assert
    let descriptionOfOutcome = '';
    if (errorString !== undefined) {
        descriptionOfOutcome = errorString;
    } else if (result === undefined) {
        descriptionOfOutcome = 'Could not find line for task';
    } else {
        descriptionOfOutcome = 'Success. Line found OK. No error reported.';
    }
    verify(descriptionOfOutcome);

    if (expectedLineNumber !== undefined) {
        expect(result).not.toBeUndefined();
        expect(result).toEqual(expectedLineNumber);

        const expectedLine = actualIncorrectLineFound ? actualIncorrectLineFound : everything.taskData.originalMarkdown;
        expect(everything.fileData.fileLines[result!]).toEqual(expectedLine);
    } else {
        expect(result).toBeUndefined();
    }
}

describe('replaceTaskWithTasks', () => {
    it('valid 2-task test', () => {
        const jsonFileName = 'single_task_valid_data.json';
        const taskLineToToggle = '- [ ] #task task 2';
        const expectedLineNumber = 5;
        testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
    });

    // --------------------------------------------------------------------------------
    // Issue 688
    describe('issue 688 - block referenced task', () => {
        const jsonFileName = '688_toggle_block_referenced_line_overwrites_wrong_line.json';
        const taskLineToToggle = '- [ ] #task task2b ^ca47c7';

        it.failing('correct behaviour', () => {
            // An incorrect line is currently found, so this test fails, due to bug 688
            const expectedLineNumber = 8;
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });

        it('current incorrect behaviour', () => {
            // An incorrect line is currently found, due to bug 688.
            // It is recognised as an incorrect line, and so line number is returned as undefined.
            const expectedLineNumber = undefined;
            // const actualIncorrectLineFound = '- [ ] #task task1a';
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });
    });

    // --------------------------------------------------------------------------------
    // Issue 1680
    describe('issue 1680 - Cannot read properties of undefined', () => {
        const jsonFileName = '1680_task_line_number_past_end_of_file.json';
        const taskLineToToggle = '- [ ] #task Section 2/Task 2';

        it.failing('correct behaviour', () => {
            // An incorrect line is currently found, so this test fails, due to bug 1680
            const expectedLineNumber = 9;
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });

        it('current incorrect behaviour', () => {
            // An incorrect line is currently found, due to bug 688.
            // It is recognised as an incorrect line, and so line number is returned as undefined.
            const expectedLineNumber = undefined;
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });
    });

    // --------------------------------------------------------------------------------
});
