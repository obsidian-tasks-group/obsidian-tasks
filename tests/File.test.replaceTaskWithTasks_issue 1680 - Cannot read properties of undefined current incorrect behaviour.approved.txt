Could not find line for task
